src/mqtt_client.rs

```diff
@@ -314,16 +309,32 @@ where
             &[Property<'a>],
         ),
     {
-        if !self.session_state.connected {
-            if let ReceivedPacket::ConnAck(acknowledge) = packet {
+        // All packets must be received in the active state.
+        if self.connection_state.state() != &States::Active {
+            // If we are establishing and just received a connack, this packet is also acceptable.
+            if !(self.connection_state.state() == &States::Establishing
+                && matches!(packet, ReceivedPacket::ConnAck(_)))
+            {
+                error!(
+                    "Received invalid packet outside of connected state: {:?}",
+                    packet
+                );
+                return Err(Error::Protocol(ProtocolError::Invalid));
+            }
+        }
+
+        match packet {
+            ReceivedPacket::ConnAck(acknowledge)
+                if self.connection_state.state() == &States::Establishing =>
```

Comment:
This is the same check as in L315. I would either move the `is_connected()` check above into the other match arms as guard. Or split the ConnAck arm out of the match. Then there is a single failure arm for unsupported packet/state combination.
Overall this is a very big function. Let's try to make it a bit more readable.